metacat-main/src/main/java/com/netflix/metacat/main/services/search/ElasticSearchMigrationUtil.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2016 Netflix, Inc.
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package com.netflix.metacat.main.services.search;
+
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.collect.Lists;
+import com.netflix.metacat.common.MetacatRequestContext;
+import com.netflix.metacat.common.json.MetacatJson;
+import com.netflix.metacat.common.server.Config;
+import org.elasticsearch.client.Client;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.util.Collections;
+import java.util.List;
+import java.util.ArrayList;
+
+/**
+ * Class for migrating the elastic search metacat index to a new one. During the migration period, metacat
+ * writes and updates the data to both the existing and the new merge index. This class overrides the
+ * save, update, and softdelete methods. For the save op, it saves to bothe indices. for the update and softdelete,
+ * it first executes the normal index ops, then copies to the merge index
+ */
+public class ElasticSearchMigrationUtil extends ElasticSearchUtil {
```

Comment:
Yes, we have been discussing this. The idea to separate them into two classes is to make the code cleaner. However, I don't see it actually bring much value given the unnecessary complexity bought into the binding phase. I think using one class is better. Also, this migration feature is rarely used. It makes sense to me to limit the 'migration' related code in one implementation class, and the upstream application should not know this.

The way to implement the migration is a bit different from writing to multiple indices using a loop. The reason is that we need to using softdelete and update, which will fail if the data were not there. Thus, for the methods save, we are saving them into the secondary indexes after saving to the original one. For the softdelete and updates, we are making the changes to the existing one, and then copy over to the secondary index. Otherwise, we will need to first copy them and then executing softdelete/update which is inefficient.